### PR TITLE
Initial support for spanning deletions and refactoring for testability

### PR DESCRIPTION
**Pull request by kcibul**
**Spanning Deletion Changes**

1. Created a new SpanningDeletionRecord as a subclass of ReferenceRecord but allows us to store GT and GQ
2. in handlePotentialSpanningDeletion when processing a deletion, we create a new SpanningDeletionRecord with the correct GT and GQ based on the deletion
3. When processing reference data at a variant site, return ReferenceRecords/SpanningDeletionRecord instead of just a string "state" since we need more than just state now
4. Because of the above, we are now returning an object for the inferred state instead of just a string.  Since the inferred state is so, so common a singleton InferredReferenceRecord was created
5. Processing of spanning deletions beyond

**Ugly**
1. The construction of the singleton is ugly because it _requires_ a location even though we don't for this case.  We could go to an tagging interface (like Cloneable) these all implement, but that seems ugly also

*Refactoring Changes*
One of the challenges with this PR was testing as the work is really done in the lower-level methods and it would be nice to have this as a unit test rather than an integration/end-to-end test.  This motivated the following changes:

1. don't write to VCF directly, instead have take a Consumer<VariantContext> to emit VariantContexts.  This let's us provide a different consumer in unit tests to collect our result.
2. we previously had a chain of calls createVariantsFromSortedRanges -> processSampleRecordsForLocation -> finalizeCurrentVariant that returned void and as a side effect wrote to VCF.  These deeper methods now return a VariantContext and the writing (via consumer) is done higher up in the call stack
3. made some private methods package-private so we could call them from tests

**Thinking Out Loud**

We have three different sets of datastructures for the same data, some of this is history, some is performance/memory, but could use some rethinking
1. GenericRecord (pulling from BQ)
2. ReferenceRecord/SpanningDeletionRecord (in memory data structure for reference data)
3. ExtractCohortRecord (pre-variant context record)

### CODE REVIEW

**Comment by kcibul** [comment]
PR at https://github.com/broadinstitute/gatk/pull/7857

**Comment by mcovarr** [comment]
I grok the refactoring changes and those are 👍, but I think I'm going to need a walkthrough for the core spanning deletion work...

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortEngine.java [hunk 1]
```diff
@@ -1120,24 +1143,103 @@ private void createVariantsFromSortedRanges(final SortedSet<Long> sampleIdsToExt
...
+                int longer = orderedDeletionLengths.get(1);  // 5
+                int lengthHomRegion = shorter;
+                int lengthHetRegion = longer - shorter;
+
+                mergeSpanningDeletionIntoCache(referenceCache.get(sample),
+                        new SpanningDeletionRecord(position + 1, sample, lengthHomRegion, "1/1", vetRow.getCallGQ())
+                );
+
+                // in the case of a single homozygous deletion, the length of the het region is zero so don't add it
+                if (lengthHetRegion > 0) {
```
[the remark this answers is not in the document]
Comment (kcibul) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortEngine.java:1187]: specifically the IF, not the contents

**Review by mcovarr — APPROVED** [review]
Thanks again for the walkthrough, I feel like I actually understood the core logic this time 😅

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortEngine.java [hunk 2]
```diff
@@ -1120,24 +1143,103 @@ private void createVariantsFromSortedRanges(final SortedSet<Long> sampleIdsToExt
...
+            if (orderedDeletionLengths.size() == 1) {
+                boolean isHet
+                        = Arrays.stream(vetRow.getCallGT().replace("|", "/").split("/")).filter(x -> !x.equals(Allele.NO_CALL_STRING)).distinct().count() == 2;
+
+                mergeSpanningDeletionIntoCache(referenceCache.get(sample),
+                        new SpanningDeletionRecord(position + 1, sample, orderedDeletionLengths.get(0), isHet ? "0/1" : "1/1", vetRow.getCallGQ())
+                );
+            } else if (orderedDeletionLengths.size() == 2) {
+                // we need to add two different spanning deletions, one for the hom region and one for the het region
+                int shorter = orderedDeletionLengths.get(0); // 5
```
Comment (gbggrant) [inline src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortEngine.java:1177]: intelliJ says that shorter is redundant with lengthHomRegion

File: src/test/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortEngineTest.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        assertVariant(vc2, "sample_1", "*/*", 42);
+        assertVariant(vc2, "sample_2", "G/T", 50);
+
+        VariantContext vc3 = results.get(3);
+        assertVariant(vc3, "sample_1", "A/*", 37);
+        assertVariant(vc3, "sample_2", "A/C", 50);
+
+        VariantContext vc4 = results.get(4);
+        assertVariant(vc4, "sample_1", "C/C", 60);
+        assertVariant(vc4, "sample_2", "C/T", 50);    }
```
Comment (gbggrant) [inline src/test/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortEngineTest.java:290]: Odd place for a `}`

**Review by gbggrant — APPROVED** [review]

**Pull request closed without merging**